src/main/java/org/apache/sysml/hops/rewrite/RewriteEMult.java

```diff
@@ -0,0 +1,284 @@
+			orderDataType = new int[dtValues.length];
+			for (int i = 0, valuesLength = dtValues.length; i < valuesLength; i++) {
+				switch(dtValues[i]) {
+				case SCALAR:
+					orderDataType[i] = 4;
+					break;
+				case MATRIX:
+					orderDataType[i] = 3;
+					break;
+				case FRAME:
+					orderDataType[i] = 2;
+					break;
+				case OBJECT:
+					orderDataType[i] = 1;
+					break;
+				case UNKNOWN:
+					orderDataType[i] = 0;
+					break;
+				}
+			}
+		}
+	};
+
+	/**
+	 * Check if a node has a parent that is not in the set of emults. Recursively check children who are also emults.
+	 * @param emults The set of BinaryOp element-wise multiply hops in the emult chain.
+	 * @param child An interior emult hop in the emult chain dag.
+	 * @return Whether this interior emult or any child emult has a foreign parent.
+	 */
+	private static boolean checkForeignParent(final Set<BinaryOp> emults, final BinaryOp child) {
+		final ArrayList<Hop> parents = child.getParent();
+		if (parents.size() > 1)
+			for (final Hop parent : parents)
+				//noinspection SuspiciousMethodCalls
+				if (!emults.contains(parent))
+					return false;
+		// child does not have foreign parents
+
+		final ArrayList<Hop> inputs = child.getInput();
+		final Hop left = inputs.get(0), right = inputs.get(1);
+		return  (!isBinaryMult(left) || checkForeignParent(emults, (BinaryOp)left)) &&
+				(!isBinaryMult(right) || checkForeignParent(emults, (BinaryOp)right));
+	}
+
+	/**
+	 * Create a set of the counts of all BinaryOp MULTs in the immediate subtree, starting with root, recursively.
+	 * @param root Root of sub-dag
+	 * @param emults Out parameter. The set of BinaryOp element-wise multiply hops in the emult chain (including root).
+	 * @param leaves Out parameter. The multiset of multiplicands in the emult chain.
+	 */
+	private static void findEMultsAndLeaves(final BinaryOp root, final Set<BinaryOp> emults, final Multiset<Hop> leaves) {
+		// Because RewriteCommonSubexpressionElimination already ran, it is safe to compare by equality.
+		emults.add(root);
+
+		final ArrayList<Hop> inputs = root.getInput();
+		final Hop left = inputs.get(0), right = inputs.get(1);
+
+		if (isBinaryMult(left)) findEMultsAndLeaves((BinaryOp) left, emults, leaves);
+		else leaves.add(left);
+
+		if (isBinaryMult(right)) findEMultsAndLeaves((BinaryOp) right, emults, leaves);
+		else leaves.add(right);
+	}
+
+	/**
+	 * Only optimize a subtree of emults if there are at least two emults.
+	 * @param emults The set of BinaryOp element-wise multiply hops in the emult chain.
+	 * @param leaves The multiset of multiplicands in the emult chain.
+	 * @return If the multiset is worth optimizing.
+	 */
+	private static boolean isOptimizable(Set<BinaryOp> emults, final Multiset<Hop> leaves) {
+		// Old criterion: there should be at least one repeated leaf
+//		for (Multiset.Entry<Hop> hopEntry : leaves.entrySet()) {
+//			if (hopEntry.getCount() > 1)
+//				return true;
+//		}
```

Comment:
Actually I like the sorting a lot because this canonicalizes these expressions. Down the road, we will likely normalize more general expressions so, please go for the current option.